Services/Common/OJS.Services.Common/Implementations/ContestsActivityService.cs

```diff
@@ -66,25 +74,46 @@ public bool CanBeCompeted(IContestForActivityServiceModel contest)
             return false;
         }
 
-        var now = this.dates.GetUtcNow();
+        var currentTimeInUtc = this.dates.GetUtcNow();
 
         if (!contest.EndTime.HasValue)
         {
             // Compete forever
-            return contest.StartTime <= now;
+            return contest.StartTime <= currentTimeInUtc;
         }
 
-        return contest.StartTime <= now && now <= contest.EndTime;
+        //If the above conditional statements are not entered, first we have to check
+        //the the start and end time > check if start is before now and end is after now
+        //if this check returns false we have to check if the current user is a participant
+        //with a remaining time in an online contest
+        return (contest.StartTime <= currentTimeInUtc && currentTimeInUtc <= contest.EndTime) ||
+               (contest.IsOnline && this.IsActiveParticipantInOnlineContest(contest.Id));
     }
 
-    public bool CanBePracticed(IContestForActivityServiceModel contest)
+    // Usage: assign value to the CanBeCompeted/Practiced properties in the different Contest models sent to the UI
+    // method must be called on model/collection after retrieving it from the db
+    // and before sending it to the UI as response
+    public void SetCanBeCompetedAndPracticed<T>(T contestModel)
+        where T : ICanBeCompetedAndPracticed
     {
-        if (!contest.IsVisible)
-        {
-            return false;
-        }
+        var contestActivity = this.GetContestActivity(contestModel.Id).GetAwaiter().GetResult();
+        contestModel.CanBeCompeted = contestActivity.CanBeCompeted;
+        contestModel.CanBePracticed = contestActivity.CanBePracticed;
+    }
 
-        if (contest.IsDeleted)
+    public bool IsActiveParticipantInOnlineContest(int contestId)
+    {
+        var currentUser = this.userProvider.GetCurrentUser();
+        var participants = this.participantsCommonData.GetAllByContest(contestId);
```

Comment:
Wouldnt it be a more optimized query if you get participants by userId not by contest?

Reply:
Yes, a single user will most like has less participants rather then all participants of contest. Refactoring it.